### PR TITLE
Fixes #1161

### LINKED ISSUE #1161: Display SEP and Equation Weighting
After [NSSService Issue 263](https://github.com/USGS-WiM/NSSServices/issues/236) is complete and
Seths weighting code is edited we need to:

- [ ] Print SEP (Si) in the report
- [ ] Send SEP (Si) to Seth's weighting code and get a response
- [ ] From the response for Seth's code display results like below:

Add a section at the bottom of the report ("Methods Weighting"): with Q (10^Z), units, PIl (Prediction Interval-Lower), PIu (Prediction Interval-Upper), and SEP
image

CI = 1.64*SEP
PIU = 10^(Z+CI)
PIL = 10^(Z-CI)

### PR DESCRIPTION
**Pull request by harper-wavra**
- src/appConfig.js - added weighting service URL and added `EquationWeighting` to MT applications.
- src/Views/sidebarview.html - added text to State/Region Specific Functions section [example](https://user-images.githubusercontent.com/59063181/164785572-db91c4ed-022f-4a9c-a147-7f23c4ff0cff.png)
- src/Views/reportview.html - added equation weighting table to report
- src/Services/nssService.ts - tried to summarize below what I did, let me know if you have questions.
    - If the selected region has `EquationWeighting` enabled and Peak-Flow Statistics are selected we will attempt to weight equations
    - Loops through all regression regions in the peak-flow statistics group. For each regression region I loop through the results and get the code, regression region name, and percent weight. If the value is greater than 0 we will send it to the equation weighting services so I then grab regression region code, change `inUse` to true, and grab the value, SEP and code (ex BWPK66_7AE) for each result.  
    - I then get the number of regression regions in the delineated basin and the number of results that need to be weighted
    - Sort the values alphabetically so they are in the same order for sending to the service
    - I then send the inputs to the `recursiveAreaWeightSubscription` function
    - Inside that function I get the code to send to the weighting service and format the json to send. It then makes that service call and once its done calls `recursiveAreaWeightSubscription` function again to make the rest of the calls
    - Once we have weighted all regression regions we check if we need to area weight the results. If we do need to area weight, we loop through all of the results and take the Z, PIl, PIu, and SEPZ value for each result and multiple by the percent weight then add them together. I then add those values to the `equationWeightingResults` to be printed out.

Testing:
If you look at the POST request for "Peak-Flow Statistics" the client makes to https://test.streamstats.usgs.gov/nssservices/scenarios/estimate?regions=MT,NA that should show all the input values, including SEP. I found this helpful to look at when checking that the correct values are sent to weighting service.

I tested these points the most:
47.41765, -111.34017 - three regression region - no SEP values are returned from the services so you cant equation weight
47.33304, -109.01935 - two regression region
47.30030, -109.08651 - one regression region

with these BC values:
WACTCH 20
WBANKFULL 17
CHANWD_RS 10

Note - A lot of time when it doesn't calculate you get the error shown like the one with point 47.41765, -111.34017. This is due to the `paramsOutOfRange` being true and the services sending back `Errors`, `EquivalentYear`, `IntervalBounds` and `SEP` back as null.

### CODE REVIEW

File: src/Views/sidebarview.html [hunk 1]
```diff
@@ -116,6 +116,7 @@ <h5>Supported search strings:</h5>
                                 <button ng-show="vm.regionService.selectedRegion.Applications.indexOf('Wateruse') != -1" ng-click="vm.OpenWateruse()" type="button" class="btn-black fullwidth"><i class="fa fa-empire" analytics-on analytics-event="Navigation" analytics-category="SideBar" analytics-label="WaterUse"></i>&nbsp;&nbsp;Check for Water Use</button>
                                 <div class="banner" ng-show="vm.regionService.selectedRegion.Applications.indexOf('KarstCheck') != -1">Check Delineated Basin for Karst Overlap (automatic)</div>
                                 <div class="banner" ng-show="vm.regionService.selectedRegion.Applications.indexOf('CoordinatedReach') != -1">Check for Coordinated Reach (automatic)</div>
+                                <div class="banner" ng-show="vm.regionService.selectedRegion.Applications.indexOf('EquationWeighting') != -1">Weight Peak-Flow Equations Based on Channel Width(automatic)</div>
```
Comment (amedenblik) [inline src/Views/sidebarview.html]: add a space before "(automatic)"

File: src/Views/reportview.html [hunk 2]
```diff
@@ -285,6 +285,38 @@ <h3 class="reportChartHeader">Estimated Flows</h3>
                         </div><!--end sc-->
                     </div><!--end extension-->
 
+                <div class="well well-sm"  ng-if="vm.regionService.selectedRegion.Applications.indexOf('EquationWeighting') != -1 && vm.nssService.selectedStatisticsGroupList.name.indexOf('Peak-Flow Statistics') != -1">
+                    <div class="wim-alert-warning" ng-show="vm.nssService.equationWeightingDisclaimers == true">Only 3 estimation methods can be weighted; the 3 estimation methods with lowest SEP values were weighted. </div>
+                    <h3>Methods Weighting</h3>
```
Comment (amedenblik) [inline src/Views/reportview.html]: Do we want a more descriptive title for this section? Maybe "Channel-width Methods Weighting [Area-Averaged]"?

File: src/Views/reportview.html [hunk 3]
```diff
@@ -285,6 +285,38 @@ <h3 class="reportChartHeader">Estimated Flows</h3>
...
+                                        <td>{{stat.Unit.abbr}}</td> 
+                                        <td>{{stat.PIl.toUSGSvalue()}}</td> 
+                                        <td>{{stat.PIu.toUSGSvalue()}}</td> 
+                                        <td>{{stat.SEPZ.toUSGSvalue()}}</td> 
+                                    </tr>
+                                </tbody>
+                            </table>
+                        </div>
+                    </div>
+                </div>
```
Comment (amedenblik) [inline src/Views/reportview.html:327]: Add citation to bottom of this section? 

_Channel-width Methods Weighting_
**[Chase, K.J., Sando, R., Armstrong, D.W., and McCarthy, P., 2021, Regional regression equations based on channel-width characteristics to estimate peak-flow frequencies at ungaged sites in Montana using peak-flow frequency data through water year 2011 (ver. 1.1, September 2021): U.S. Geological Survey Scientific Investigations Report 2020–5142, 49 p.](https://doi.org/10.3133/sir20205142)**

File: src/Views/reportview.html [hunk 4]
```diff
@@ -285,6 +285,38 @@ <h3 class="reportChartHeader">Estimated Flows</h3>
                         </div><!--end sc-->
                     </div><!--end extension-->
 
+                <div class="well well-sm"  ng-if="vm.regionService.selectedRegion.Applications.indexOf('EquationWeighting') != -1 && vm.nssService.selectedStatisticsGroupList.name.indexOf('Peak-Flow Statistics') != -1">
```
Comment (amedenblik) [inline src/Views/reportview.html]: Add collapsibility to this section, like in #1281?

File: src/Views/reportview.html [hunk 5]
```diff
@@ -285,6 +285,38 @@ <h3 class="reportChartHeader">Estimated Flows</h3>
                         </div><!--end sc-->
                     </div><!--end extension-->
 
+                <div class="well well-sm"  ng-if="vm.regionService.selectedRegion.Applications.indexOf('EquationWeighting') != -1 && vm.nssService.selectedStatisticsGroupList.name.indexOf('Peak-Flow Statistics') != -1">
+                    <div class="wim-alert-warning" ng-show="vm.nssService.equationWeightingDisclaimers == true">Only 3 estimation methods can be weighted; the 3 estimation methods with lowest SEP values were weighted. </div>
+                    <h3>Methods Weighting</h3>
+                    <div ng-if="vm.nssService.equationWeightingResults.length == 0">No method weighting results returned</div>
```
Comment (amedenblik) [inline src/Views/reportview.html]: Add period here: "No method weighting results returned."

File: src/Views/reportview.html [hunk 6]
```diff
@@ -285,6 +285,38 @@ <h3 class="reportChartHeader">Estimated Flows</h3>
                         </div><!--end sc-->
                     </div><!--end extension-->
 
+                <div class="well well-sm"  ng-if="vm.regionService.selectedRegion.Applications.indexOf('EquationWeighting') != -1 && vm.nssService.selectedStatisticsGroupList.name.indexOf('Peak-Flow Statistics') != -1">
+                    <div class="wim-alert-warning" ng-show="vm.nssService.equationWeightingDisclaimers == true">Only 3 estimation methods can be weighted; the 3 estimation methods with lowest SEP values were weighted. </div>
+                    <h3>Methods Weighting</h3>
+                    <div ng-if="vm.nssService.equationWeightingResults.length == 0">No method weighting results returned</div>
+                    <div ng-if="vm.nssService.equationWeightingResults.length != 0">
```
Comment (amedenblik) [inline src/Views/reportview.html]: Before this section, include definitions (similar to other sections of the report):

"PIl: Prediction Interval-Lower, PIu: Prediction Interval-Upper, ASEp: Average Standard Error of Prediction"

File: src/appConfig.js [hunk 7]
```diff
@@ -374,7 +375,7 @@ configuration.regions = [
                     "queryProperties": { "Regulation Points": { "Descript": "Description" } }
                 }
             },
-        "Applications": ["Regulation"],
+        "Applications": ["Regulation", "EquationWeighting"],
```
Comment (amedenblik) [inline src/appConfig.js]: Should we rename to something more specific, like ChannelWidthWeighting? Not sure if other types of equation weighting may come up in the future.

File: src/Views/reportview.html [hunk 8]
```diff
@@ -285,6 +285,38 @@ <h3 class="reportChartHeader">Estimated Flows</h3>
                         </div><!--end sc-->
                     </div><!--end extension-->
 
+                <div class="well well-sm"  ng-if="vm.regionService.selectedRegion.Applications.indexOf('EquationWeighting') != -1 && vm.nssService.selectedStatisticsGroupList.name.indexOf('Peak-Flow Statistics') != -1">
+                    <div class="wim-alert-warning" ng-show="vm.nssService.equationWeightingDisclaimers == true">Only 3 estimation methods can be weighted; the 3 estimation methods with lowest SEP values were weighted. </div>
+                    <h3>Methods Weighting</h3>
+                    <div ng-if="vm.nssService.equationWeightingResults.length == 0">No method weighting results returned</div>
+                    <div ng-if="vm.nssService.equationWeightingResults.length != 0">
+                        <div ng-repeat="rr in vm.nssService.equationWeightingResults">
+                            <h4>{{rr.RR}}</h4>
```
Comment (amedenblik) [inline src/Views/reportview.html]: Should Regression Region be listed here, if the results are area-averaged?

**Review by amedenblik — CHANGES_REQUESTED** [review]
Impressive work! It is working well. Please see code comments. Two additional comments:

- [x] Example of potential error with this input: https://ss-weightingservices.streamstats.usgs.gov/weightest/

POST body: {"x1":178,"x2":52.5,"x3":68.6,"x4":49.6,"sep1":0.23220407045532923,"sep2":0.3880909941793581,"sep3":0.2759430640218844,"sep4":0.44902115763068445,"regressionRegionCode":"GC1849","code1":"ACPK50AEP","code2":"PK50AEP","code3":"BWPK50AEP","code4":"RSPK50AEP"}

Warning message returned: "Weighted value is outside the range of input values. Only 3 estimation methods can be weighted; the 3 estimation methods with lowest SEP values were weighted." 

The part about "Weighted value is outside the range of input values" is not provided in the Report. For the test point I was looking at (46.96753, -109.12239), three of the statistics returned this warning message.

- [x] Methods Weighting section is not provided when you Download the report as a CSV. - added issue [#1293](https://github.com/USGS-WiM/StreamStats/issues/1293)
- [x] Report opens before timeout for the regression region results and area-averaged results finishes, so you need to click on report to "refresh" and to see results. 

Notes:
- ~~I created https://github.com/USGS-WiM/SS-WeightingServices/issues/19 to improve error handling for valid x values with invalid SEP values.~~ this has been incorporated
- After merging, we should have Pete & Montana team review for accuracy.

File: src/Views/reportview.html (hunk 8 above)
Comment (harper-wavra) [reply src/Views/reportview.html]: I guess I didn't do to keep it similar to the statistics section of the report, we just have it listed as Area-Averaged in that section.
image

Comment (amedenblik) [reply src/Views/reportview.html]: Turns out this is an issue related to timeouts for the regression region results and area-averaged results.

**Review by amedenblik — CHANGES_REQUESTED** [review]
- [x] Change toast error that says "Cannot Methods Weight" to "Cannot Weight Channel-width Methods" to be more clear 

Otherwise looks great!

**Review by amedenblik — APPROVED** [review]

**Pull request merged**